File: python/etsm/test/etsm_test.py
import unittest
import etsm

test = None

class FooState(etsm.State):
    def __init__(self, enter, exit, tick):
        super().__init__(enter, exit)
        self.Tick = tick

class Foo:
    def __init__(self):
        self.__sm = etsm.StateMachine(self)
        self.__a = FooState(Foo.EnterA, Foo.ExitA, Foo.TickA)
        self.__b = FooState(Foo.EnterB, None, Foo.TickB)
        self.__output = ''

    def EnterA(self):
        self.__output += ' ->A '
        global test
        test.assertEqual( self.__sm.Transition(self.__a), False)

    def ExitA(self):
        self.__output += ' A-> '
        global test
        test.assertEqual( self.__sm.Transition(self.__a), False)

    def TickA(self):
        self.__output += ' A '

    def EnterB(self):
        self.__output += ' ->B '
        global test
        test.assertEqual( self.__sm.Transition(self.__a), False)

    # no exit for B
    # def ExitB(self):
    #     self.__output += ' B-> '
    #     global test
    #     test.assertEqual( self.__sm.Transition(self.__a), False)

    def TickB(self):
        self.__output += ' B '

    # call tick on the current state, aka virtual call
    def Tick(self):
        if ( self.__sm.Current != None ):
            self.__sm.Current.Tick(self)

    def Test(self):
        global test
        self.Tick();
        self.__sm.Transition(self.__a);
        test.assertEqual(self.__sm.IsIn(self.__a), True)
        self.Tick();
        self.__sm.Transition(self.__b);
        test.assertEqual(self.__sm.IsIn(self.__b), True)
        self.Tick();
        self.__sm.Transition(None);
        self.Tick();
        test.assertEqual(self.__output, ' ->A  A  A->  ->B  B ')

class TestMethods(unittest.TestCase):

    def test_sample(self):
        global test
        test = self
        foo = Foo()
        foo.Test()

if __name__ == '__main__':
    unittest.main()

File: c++/test/etsm_test.cpp
#include <etsm.h>

#include <iostream>
#include <assert.h>
#include <string>

using namespace etsm;

namespace test_ab
{
    class Foo
    {
    public:
        Foo()
            : sm(this),
              a(&Foo::EnterA, &Foo::ExitA),
              b(&Foo::EnterB, nullptr)
        {
        }

        void Test()
        {
            assert(sm.IsIn(nullptr));
            sm.Transition(&a);
            assert(sm.IsIn(&a));
            sm.Transition(&b);
            assert(sm.IsIn(&b));
            sm.Transition(nullptr);
            assert(sm.IsIn(nullptr));
            assert(output == " ->A  A->  ->B ");
        }

    private:
        StateMachine<Foo> sm;
        State<Foo> a;
        State<Foo> b;
        std::string output;

        void EnterA() { output += " ->A "; }
        void ExitA() { output += " A-> "; }
        void EnterB() { output += " ->B "; }
    };
}

namespace test_virtual_call
{
    class Foo;

    class FooState : public State<Foo>
    {
    public:
        FooState(Method enter, Method exit, Method tick = nullptr)
            : State<Foo>(enter, exit),
              tick(tick)
        {}

        void Tick(Foo* owner)
        {
            if (tick != nullptr)
                (owner->*tick)();
        }

    private:
        Method tick;
    };

    class Foo
    {
    public:
        Foo()
            : sm(this),
            a(nullptr, nullptr, &Foo::TickA),
            b(nullptr, nullptr, &Foo::TickB)
        {}

        void Tick()
        {
            if (sm.GetCurrent())
                sm.GetCurrent()->Tick(this);
        }

        void Test()
        {
            Tick();
            sm.Transition(&a);
            Tick();
            sm.Transition(&b);
            Tick();
            sm.Transition(nullptr);
            assert(output == " A  B ");
        }

    private:
        StateMachine<Foo, FooState> sm;
        FooState a;
        FooState b;
        std::string output;

        void TickA() { output += " A "; }
        void TickB() { output += " B "; }
    };
}

int main()
{
    test_ab::Foo().Test();
    test_virtual_call::Foo().Test();
}